File: parent/84 -- Largest Rectangle in Histogram /mine2.py
class Solution:
    def largestRectangleArea(self, heights: List[int]) -> int:
        # Traditional Monotonic Stack approach. More optimal for space
        # Time: O(n)
        # Space: O(n)

        n = len(heights)
        stack = [(0, heights[0])]     # (idx, height)
        areas = []
        i = 1
        res = 0

        while i < n:
            if heights[i] >= heights[i-1]:
                stack.append((i, heights[i]))
                i += 1
                continue
            
            idx = 0
            while stack and stack[-1][1] >= heights[i]:
                idx, num = stack.pop()
                res = max(res, num * (i - idx))
            
            stack.append((idx, heights[i]))
            i += 1
        
        for idx, num in stack:
            res = max(res, num * (n - idx))
        
        return res

File: parent/84 -- Largest Rectangle in Histogram /mine2.go
package main

func largestRectangleArea(heights []int) int {
	n := len(heights)
	var res int
	stack := [][2]int{[2]int{0, heights[0]}} // (idx, num)

	i := 1

	for i < n {
		if heights[i] >= heights[i-1] {
			stack = append(stack, [2]int{i, heights[i]})
			i += 1
			continue
		}

		idx := 0
		for len(stack) > 0 && heights[i] < stack[len(stack)-1][1] {
			item := stack[len(stack)-1]
			idx = item[0]
			val := item[1]
			stack = stack[:len(stack)-1]

			res = max(res, val*(i-idx))
		}
		stack = append(stack, [2]int{idx, heights[i]})
		i += 1
	}
	for _, item := range stack {
		idx, val := item[0], item[1]
		res = max(res, val*(n-idx))
	}

	return res
}
